include/awkward/Content.h

```diff
@@ -1184,6 +1184,23 @@ namespace awkward {
     static int64_t
       axis_wrap_if_negative(int64_t axis);
 
+    /// @brief Transfer the entire contents of the array on to the GPU.
+    ///
+    /// Returns a std::shared_ptr<Content> which is, by default, allocated on the
+    /// first device(device [0])
+    ///
+    /// @note This function has not been implemented to handle Multi-GPU setups
+    virtual ContentPtr
+      to_gpu(kernel::Lib ptr_lib) const;
+
+    /// @brief Transfer the contents of the array on the GPU to the main memory
+    ///
+    /// Returns a std::shared_ptr<Content> which is, allocated on the main memory
+    ///
+    /// @note This function has not been implemented to handle Multi-GPU setups
+    virtual ContentPtr
+      to_cpu() const;
+
```

Comment:
Okay so, the `to_cpu` internally runs itself through all the kernels except itself. (It reads the class variable internally `ptr_lib_` and dispatches the appropriate kernel according to `ptr_lib_`). Although, I can manage this here, There would be a mismatch when is use the kernel dispatcher. It would be something like just `copy_to(kernel::Lib::cuda_kernel)`. 
But you can have two options here, either to `copy_from` or `copy_to`. Since the main awkward1 runs on main memory, wouldn't it be useful to have two functions `copy_from` and `copy_to` and it would be intuitive to think along those lines too, given that `main_memory` is the subject of discussion. This should also address the kernel naming that you mentioned below. What do you think?

Reply:
A "copy to" function makes more sense than a "copy from" function, and it should take either CPU or CUDA pointers as arguments. (It's a pass-through, changing nothing, if the buffer is already in the right place.) It will often be the case that a user starts with an array that they don't know (or care) whether it's currently in main memory or a GPU, but they want to make it one or the other—especially if there are more than two options (the hypothetical "AMD" case).

Reply:
I just pushed the changes concerning this and I can't help but thank you for urging me to go this way. The entire code base is much cleaner know, and more importantly, much more maintainable. So now, for any future libraries, just adding the corresponding kernel in the `kernel` namespace would be suffice. We would never have to touch the `copy_to` function again.

Reply:
Great! `:)`